### PR TITLE
Add torch.no_grad decorator for saving gpu mem

### PR DESCRIPTION
**Pull request by triple-Mu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation for this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by hhaAndroid — APPROVED** [review]

**Comment by hhaAndroid** [comment]
@triple-Mu Is there a gradient in the export phase? My understanding is that this decorator should not be needed?

**Comment by triple-Mu** [comment]
If with out `torch.no_grad` :
``` text
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    78    493.5 MiB    493.5 MiB           1   @profile
    79                                         # @torch.no_grad()
    80                                         def main():
    81    493.5 MiB      0.0 MiB           1       args = parse_args()
    82    494.0 MiB      0.4 MiB           1       register_all_modules()
    83                                         
    84    494.0 MiB      0.0 MiB           1       mkdir_or_exist(args.work_dir)
    85                                         
    86    494.0 MiB      0.0 MiB           1       if args.model_only:
    87                                                 postprocess_cfg = None
    88                                                 output_names = None
    89                                             else:
    90    494.0 MiB      0.0 MiB           2           postprocess_cfg = ConfigDict(
    91    494.0 MiB      0.0 MiB           1               pre_top_k=args.pre_topk,
    92    494.0 MiB      0.0 MiB           1               keep_top_k=args.keep_topk,
    93    494.0 MiB      0.0 MiB           1               iou_threshold=args.iou_threshold,
    94    494.0 MiB      0.0 MiB           1               score_threshold=args.score_threshold,
    95    494.0 MiB      0.0 MiB           1               backend=args.backend)
    96    494.0 MiB      0.0 MiB           1           output_names = ['num_dets', 'boxes', 'scores', 'labels']
    97   1948.4 MiB   1454.4 MiB           1       baseModel = build_model_from_cfg(args.config, args.checkpoint, args.device)
    98                                         
    99   1948.4 MiB      0.0 MiB           2       deploy_model = DeployModel(
   100   1948.4 MiB      0.0 MiB           1           baseModel=baseModel, postprocess_cfg=postprocess_cfg)
   101   1948.4 MiB      0.0 MiB           1       deploy_model.eval()
   102                                         
   103   1952.3 MiB      4.0 MiB           3       fake_input = torch.randn(args.batch_size, 3,
   104   1952.3 MiB      0.0 MiB           2                                *args.img_size).to(args.device)
   105                                             # dry run
   106   3655.2 MiB   1702.9 MiB           1       deploy_model(fake_input)
   107                                         
   108   3655.2 MiB      0.0 MiB           1       save_onnx_path = os.path.join(args.work_dir, 'end2end.onnx')
   109                                             # export onnx
   110   3655.2 MiB      0.0 MiB           1       with BytesIO() as f:
   111   3764.6 MiB    109.3 MiB           2           torch.onnx.export(
   112   3655.2 MiB      0.0 MiB           1               deploy_model,
   113   3655.2 MiB      0.0 MiB           1               fake_input,
   114   3655.2 MiB      0.0 MiB           1               f,
   115   3655.2 MiB      0.0 MiB           1               input_names=['images'],
   116   3655.2 MiB      0.0 MiB           1               output_names=output_names,
   117   3655.2 MiB      0.0 MiB           1               opset_version=args.opset)
   118   3764.6 MiB      0.0 MiB           1           f.seek(0)
   119   3797.1 MiB     32.5 MiB           1           onnx_model = onnx.load(f)
   120   3834.6 MiB     37.5 MiB           1           onnx.checker.check_model(onnx_model)
   121                                         
   122                                                 # Fix tensorrt onnx output shape, just for view
   123   3834.6 MiB      0.0 MiB           1           if args.backend in (2, 3):
   124   3834.6 MiB      0.0 MiB           1               shapes = [
   125   3834.6 MiB      0.0 MiB           1                   args.batch_size, 1, args.batch_size, args.keep_topk, 4,
   126   3834.6 MiB      0.0 MiB           1                   args.batch_size, args.keep_topk, args.batch_size,
   127   3834.6 MiB      0.0 MiB           1                   args.keep_topk
   128                                                     ]
   129   3834.6 MiB    -37.3 MiB           5               for i in onnx_model.graph.output:
   130   3834.6 MiB      0.0 MiB          13                   for j in i.type.tensor_type.shape.dim:
   131   3834.6 MiB      0.0 MiB           9                       j.dim_param = str(shapes.pop(0))
   132   3797.3 MiB    -37.3 MiB           1       if args.simplify:
   133   3797.3 MiB      0.0 MiB           1           try:
   134   3800.5 MiB      3.2 MiB           1               import onnxsim
   135   4036.9 MiB    236.4 MiB           1               onnx_model, check = onnxsim.simplify(onnx_model)
   136   4036.9 MiB      0.0 MiB           1               assert check, 'assert check failed'
   137                                                 except Exception as e:
   138                                                     print(f'Simplify failure: {e}')
   139   4036.9 MiB      0.0 MiB           1       onnx.save(onnx_model, save_onnx_path)
   140   4036.9 MiB      0.0 MiB           1       print(f'ONNX export success, save into {save_onnx_path}')
```
else: 
``` text
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    78    493.8 MiB    493.8 MiB           1   @torch.no_grad()
    79                                         @profile
    80                                         def main():
    81    493.8 MiB      0.0 MiB           1       args = parse_args()
    82    494.2 MiB      0.5 MiB           1       register_all_modules()
    83                                         
    84    494.2 MiB      0.0 MiB           1       mkdir_or_exist(args.work_dir)
    85                                         
    86    494.2 MiB      0.0 MiB           1       if args.model_only:
    87                                                 postprocess_cfg = None
    88                                                 output_names = None
    89                                             else:
    90    494.2 MiB      0.0 MiB           2           postprocess_cfg = ConfigDict(
    91    494.2 MiB      0.0 MiB           1               pre_top_k=args.pre_topk,
    92    494.2 MiB      0.0 MiB           1               keep_top_k=args.keep_topk,
    93    494.2 MiB      0.0 MiB           1               iou_threshold=args.iou_threshold,
    94    494.2 MiB      0.0 MiB           1               score_threshold=args.score_threshold,
    95    494.2 MiB      0.0 MiB           1               backend=args.backend)
    96    494.2 MiB      0.0 MiB           1           output_names = ['num_dets', 'boxes', 'scores', 'labels']
    97   1951.8 MiB   1457.5 MiB           1       baseModel = build_model_from_cfg(args.config, args.checkpoint, args.device)
    98                                         
    99   1951.8 MiB      0.0 MiB           2       deploy_model = DeployModel(
   100   1951.8 MiB      0.0 MiB           1           baseModel=baseModel, postprocess_cfg=postprocess_cfg)
   101   1951.8 MiB      0.0 MiB           1       deploy_model.eval()
   102                                         
   103   1951.8 MiB      0.0 MiB           3       fake_input = torch.randn(args.batch_size, 3,
   104   1951.8 MiB      0.0 MiB           2                                *args.img_size).to(args.device)
   105                                             # dry run
   106   3659.4 MiB   1707.6 MiB           1       deploy_model(fake_input)
   107                                         
   108   3659.4 MiB      0.0 MiB           1       save_onnx_path = os.path.join(args.work_dir, 'end2end.onnx')
   109                                             # export onnx
   110   3659.4 MiB      0.0 MiB           1       with BytesIO() as f:
   111   3760.6 MiB    101.1 MiB           2           torch.onnx.export(
   112   3659.4 MiB      0.0 MiB           1               deploy_model,
   113   3659.4 MiB      0.0 MiB           1               fake_input,
   114   3659.4 MiB      0.0 MiB           1               f,
   115   3659.4 MiB      0.0 MiB           1               input_names=['images'],
   116   3659.4 MiB      0.0 MiB           1               output_names=output_names,
   117   3659.4 MiB      0.0 MiB           1               opset_version=args.opset)
   118   3760.6 MiB      0.0 MiB           1           f.seek(0)
   119   3760.6 MiB      0.1 MiB           1           onnx_model = onnx.load(f)
   120   3798.6 MiB     38.0 MiB           1           onnx.checker.check_model(onnx_model)
   121                                         
   122                                                 # Fix tensorrt onnx output shape, just for view
   123   3798.6 MiB      0.0 MiB           1           if args.backend in (2, 3):
   124   3798.6 MiB      0.0 MiB           1               shapes = [
   125   3798.6 MiB      0.0 MiB           1                   args.batch_size, 1, args.batch_size, args.keep_topk, 4,
   126   3798.6 MiB      0.0 MiB           1                   args.batch_size, args.keep_topk, args.batch_size,
   127   3798.6 MiB      0.0 MiB           1                   args.keep_topk
   128                                                     ]
   129   3798.6 MiB    -38.5 MiB           5               for i in onnx_model.graph.output:
   130   3798.6 MiB      0.0 MiB          13                   for j in i.type.tensor_type.shape.dim:
   131   3798.6 MiB      0.0 MiB           9                       j.dim_param = str(shapes.pop(0))
   132   3760.2 MiB    -38.5 MiB           1       if args.simplify:
   133   3760.2 MiB      0.0 MiB           1           try:
   134   3763.3 MiB      3.2 MiB           1               import onnxsim
   135   4038.5 MiB    275.1 MiB           1               onnx_model, check = onnxsim.simplify(onnx_model)
   136   4038.5 MiB      0.0 MiB           1               assert check, 'assert check failed'
   137                                                 except Exception as e:
   138                                                     print(f'Simplify failure: {e}')
   139   4038.5 MiB      0.0 MiB           1       onnx.save(onnx_model, save_onnx_path)
   140   4038.5 MiB      0.0 MiB           1       print(f'ONNX export success, save into {save_onnx_path}')
```
